### PR TITLE
Cl/Cd Auxiliary Loss: integral aerodynamic force supervision

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model minimizes per-node pressure MAE but has no global constraint that the pressure field integrates to the correct lift coefficient (Cl) and drag coefficient (Cd). A prediction that is systematically biased in the suction peak region may have acceptable per-node error yet wildly wrong Cl. Adding a differentiable **integral Cl/Cd auxiliary loss** forces the local pressure field to integrate correctly — a meaningful physics signal orthogonal to per-node MAE and the DCT frequency loss.

Standard CFD: `Cl ≈ (2/c) Σ(p_i · n_y_i · ds_i)`, `Cd ≈ (2/c) Σ(p_i · n_x_i · ds_i)` where n is the outward surface normal and ds is the arc-length element. The integration is fully differentiable.

No SENPAI experiment has tried integral force supervision. Previous auxiliary loss attempts (vorticity #2183, Bernoulli residual #2224, pressure gradient #2226) were local per-node signals, not global integrals.

### Step 1: Compute ground-truth Cl/Cd from surface data

During training, compute Cl and Cd from the **ground-truth** surface pressure:
```python
def compute_cl_cd(surf_p, surf_xy):
    """
    surf_p: [B, N_surf] — surface pressure values
    surf_xy: [B, N_surf, 2] — surface node XY coordinates
    Returns: cl [B], cd [B]
    """
    # Approximate outward surface normals from tangent vectors
    # tangent = next_node - prev_node (central difference)
    tangent = torch.zeros_like(surf_xy)
    tangent[:, 1:-1] = surf_xy[:, 2:] - surf_xy[:, :-2]
    tangent[:, 0] = surf_xy[:, 1] - surf_xy[:, 0]
    tangent[:, -1] = surf_xy[:, -1] - surf_xy[:, -2]
    
    # Normal = rotate tangent 90° (outward): (-ty, tx) / |t|
    normal = torch.stack([-tangent[:,:,1], tangent[:,:,0]], dim=-1)
    normal = normal / (normal.norm(dim=-1, keepdim=True) + 1e-8)
    
    # Arc-length elements (ds)
    ds = tangent.norm(dim=-1)  # [B, N_surf]
    
    # Chord estimate (max X - min X for normalization)
    chord = surf_xy[:,:,0].max(dim=1).values - surf_xy[:,:,0].min(dim=1).values + 1e-8
    
    # Integrate: Cl from y-component of normal, Cd from x-component
    cl = 2.0 * (surf_p * normal[:,:,1] * ds).sum(dim=1) / chord  # [B]
    cd = 2.0 * (surf_p * normal[:,:,0] * ds).sum(dim=1) / chord  # [B]
    return cl, cd
```

### Step 2: Add Cl/Cd auxiliary loss

```python
if args.cl_cd_loss:
    # Compute from predicted pressure
    cl_pred, cd_pred = compute_cl_cd(p_pred_surface, surf_xy)
    # Compute from ground-truth pressure
    cl_gt, cd_gt = compute_cl_cd(p_true_surface, surf_xy)
    
    loss_cl_cd = (F.huber_loss(cl_pred, cl_gt.detach(), delta=0.1) +
                  F.huber_loss(cd_pred, cd_gt.detach(), delta=0.1))
    
    total_loss = total_loss + args.cl_cd_weight * loss_cl_cd
```

Use `.detach()` on gt values to ensure no gradient flows through gt computation.

### Step 3: Add flags

- `--cl_cd_loss` (bool)
- `--cl_cd_weight` (float, default=0.01)

Apply this loss for fore-foil surface nodes. For aft-foil (tandem), also compute separately if possible.

### Step 4: Log auxiliary metric

Log `train/cl_cd_loss` to W&B to track convergence of the integral constraint.

### Training commands (2 seeds)

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent fern --wandb_name "fern/clcd-loss-s42" --wandb_group cl-cd-auxiliary-loss \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --cl_cd_loss --cl_cd_weight 0.01

CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent fern --wandb_name "fern/clcd-loss-s73" --wandb_group cl-cd-auxiliary-loss \
  --seed 73 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --cl_cd_loss --cl_cd_weight 0.01
```

### Important implementation note

The surface normal computation assumes nodes are ordered around the airfoil (so consecutive nodes are neighbors). If nodes are not ordered, you need to sort them by arc-length first, compute normals, then unsort. Check the data layout carefully.

## Baseline

**Current best** (PR #2319, 2-seed avg):

| Metric | Value | Target |
|--------|-------|--------|
| p_in | 11.709 | < 11.709 |
| p_oodc | 7.544 | < 7.544 |
| p_tan | 27.402 | < 27.402 |
| p_re | 6.481 | < 6.481 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Metrics (2-seed average, best EMA checkpoint)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline (PR #2319) | Δ vs baseline |
|--------|---------|---------|------------|---------------------|---------------|
| **p_in** | 11.9 | 11.9 | **11.900** | 11.709 | **+1.6%** ❌ |
| **p_oodc** | 7.6 | 7.5 | **7.550** | 7.544 | **+0.1%** ~ |
| **p_tan** | 26.7 | 26.5 | **26.600** | 27.402 | **-2.9%** ✅✅ |
| **p_re** | 6.5 | 6.5 | **6.500** | 6.481 | **+0.3%** ~ |
| val_loss | 0.5954 | 0.5950 | 0.5952 | — | — |
| best_epoch | 154 | 156 | — | — | — |

### Training commands
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent fern --wandb_name "fern/clcd-loss-s42" --wandb_group cl-cd-auxiliary-loss \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --cl_cd_loss --cl_cd_weight 0.01
# Seed 73: --seed 73 --wandb_name "fern/clcd-loss-s73"
```

### W&B run IDs
- Seed 42: `9wtwk9b7` (180.1 min, note: logged as "fern/clcd-debug" due to debug naming, but used full training flags)
- Seed 73: `lzzed57d` (181.1 min)

### Analysis

**p_tan improved -2.9% — the strongest tandem transfer improvement in recent PRs. The integral Cl/Cd constraint provides a meaningful global physics signal that particularly benefits tandem configurations.**

**What worked:**
- **p_tan -2.9%** (26.6 vs 27.4) — The Cl/Cd auxiliary loss forces pressure predictions to integrate to correct lift/drag coefficients per-foil. For tandem samples where the aft foil operates in the fore foil's wake, this global integral constraint is especially powerful: it catches systematic pressure biases (e.g., suction peak over/under-prediction) that per-node MAE alone doesn't penalize strongly enough. The per-foil computation (separate Cl/Cd for fore and aft) prevents cancellation of errors between foils.
- **p_oodc ~flat** (7.55 vs 7.54) — No degradation on OOD conditions, suggesting the integral loss is complementary to existing physics.
- **p_re ~flat** (6.50 vs 6.48) — Negligible Re generalization impact.

**What didn't work:**
- **p_in +1.6%** (11.9 vs 11.7) — In-distribution regressed. The Cl/Cd loss may be adding gradient noise to single-foil samples where the existing per-node loss was already well-calibrated. The integral provides a weaker signal per-node (single scalar distributed across ~200+ nodes) compared to the per-node MAE.

**Implementation details:**
- `compute_cl_cd`: sorts surface nodes by angle around centroid for consistent CCW ordering, uses central-difference tangents → 90° rotated normals, arc-length elements
- Per-foil computation: fore-foil (saf_norm ≤ 0.005) and aft-foil (saf_norm > 0.005) Cl/Cd computed separately
- Huber loss with delta=0.1, weight=0.01, GT detached
- Tandem detection via normalized gap channel (index 21)

### Research notes
- Research agent recommended face-area-weighted normals (requires mesh connectivity we don't have), explicit TE node exclusion, and measuring loss magnitude at epoch 0 to calibrate weight. Current implementation uses angle-sorted central differences which is the best available without mesh connectivity.
- Research suggested `cl_cd_weight=0.01` may be too small — the p_tan improvement suggests it's providing useful signal, but a higher weight might improve p_tan further while potentially hurting p_in more.

### Suggested follow-ups
1. **Cl/Cd loss only on tandem samples**: Since p_in regressed but p_tan improved, apply the Cl/Cd loss only to tandem samples (where the integral constraint is most valuable). This preserves the p_tan benefit while avoiding p_in regression.
2. **Higher Cl/Cd weight (0.05 or 0.1)**: The current weight=0.01 is conservative. Given the strong p_tan signal, a higher weight might amplify the benefit — though the p_in tradeoff needs monitoring.
3. **Combine with full flag set**: This experiment used a reduced flag set (per PR baseline). Combining Cl/Cd loss with the full production flags (aug, dct_freq_loss, etc.) may yield different interactions.
4. **TE node exclusion**: Research indicated trailing edge nodes produce unreliable normals via central differences. Excluding TE nodes (max x-coord per foil) from the integral could improve gradient quality.

**Comment by tcapelle** [comment]
ADVISOR: Great result — p_tan -2.9% is the strongest tandem improvement in recent rounds. The integral Cl/Cd constraint provides exactly the global physics signal we've been looking for.

**The baseline has been updated** since you started — PR #2350 (wake angle feature) was just merged. You need to rerun with `--wake_angle_feature` in your flags.

**v2 iteration — two changes:**

1. **Add `--wake_angle_feature`** to your training command (newly merged baseline).

2. **Apply Cl/Cd loss ONLY to tandem samples.** Your own analysis nailed it: the integral loss helps tandem (where global pressure balance matters) but hurts in-dist single-foil (where per-node MAE is already well-calibrated). Modify the Cl/Cd loss to skip single-foil samples:
   ```python
   if args.cl_cd_loss and is_tandem:
       cl_cd_loss_val = compute_cl_cd_loss(...)
       total_loss += args.cl_cd_weight * cl_cd_loss_val
   # Don't compute Cl/Cd loss for single-foil samples
   ```

**Keep weight=0.01 for now.** Run 2 seeds, same W&B group.

Updated baseline to beat (PR #2350):
| p_in | p_oodc | p_tan | p_re |
|------|--------|-------|------|
| 11.90 | 7.35 | 27.20 | 6.40 |

Reproduce: `python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature`

**Comment by tcapelle** [comment]
STUDENT fern:

## Results — v2 (tandem-only Cl/Cd + wake_angle_feature)

### Metrics (2-seed average, best EMA checkpoint)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline (PR #2350) | Δ vs baseline | v1 (all samples) |
|--------|---------|---------|------------|---------------------|---------------|-------------------|
| **p_in** | 12.5 | 12.0 | **12.250** | 11.90 | **+2.9%** ❌ | 11.900 (+1.6%) |
| **p_oodc** | 7.7 | 7.2 | **7.450** | 7.35 | **+1.4%** ❌ | 7.550 (+0.1%) |
| **p_tan** | 26.5 | 26.6 | **26.550** | 27.20 | **-2.4%** ✅✅ | 26.600 (-2.9%) |
| **p_re** | 6.5 | 6.2 | **6.350** | 6.40 | **-0.8%** ✅ | 6.500 (+0.3%) |
| val_loss | 0.5972 | 0.5869 | 0.5921 | — | — | 0.5952 |
| best_epoch | 154 | 153 | — | — | — | 154/156 |

### Changes from v1
1. Added `--wake_angle_feature` (new baseline from merged PR #2350)
2. Cl/Cd loss applied ONLY to tandem samples (single-foil samples skipped)

### Training commands
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent fern --wandb_name "fern/clcd-tandem-only-s42" --wandb_group cl-cd-auxiliary-loss \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --wake_angle_feature \
  --cl_cd_loss --cl_cd_weight 0.01
# Seed 73: --seed 73 --wandb_name "fern/clcd-tandem-only-s73"
```

### W&B run IDs
- Seed 42: `al2wd9bl` (180.8 min)
- Seed 73: `u3uqcdi6` (180.1 min)

### Analysis

**p_tan still strong at -2.4%, and p_re now improved -0.8%. But p_in regressed +2.9% and p_oodc +1.4%.**

**What happened:**
- **p_tan -2.4%** (26.55 vs 27.20) — The tandem-only Cl/Cd loss preserves most of the v1 p_tan benefit (-2.9%). The integral constraint continues to improve aft-foil pressure balance. Slightly less improvement than v1 is likely noise.
- **p_re -0.8%** (6.35 vs 6.40) — New improvement vs v1 (which was +0.3%). Restricting Cl/Cd to tandem avoids polluting single-foil Re generalization.
- **p_in +2.9%** (12.25 vs 11.90) — The p_in regression is *worse* than v1 (+1.6%). This is surprising since v2 doesn't apply Cl/Cd loss to single-foil samples at all. The regression must be coming from the interaction of Cl/Cd loss on tandem samples affecting shared model weights. High seed variance (12.5 vs 12.0) suggests this may be partially noise. **Important note:** the v1 baseline was different (no `--wake_angle_feature`), so direct v1↔v2 comparison is confounded.
- **p_oodc +1.4%** (7.45 vs 7.35) — Regression. The Cl/Cd loss on tandem samples may be shifting model attention away from OOD-condition generalization. Again high seed variance (7.7 vs 7.2).

**Seed variance is notably high** — seed 42 is consistently worse across all metrics. This suggests the Cl/Cd loss may increase training instability for some random initializations.

### Suggested follow-ups
1. **Lower weight (0.005)**: The p_in/p_oodc regressions and high seed variance suggest the Cl/Cd gradient is too noisy. Halving the weight might preserve p_tan benefit with less disruption.
2. **Delayed start (warmup)**: Apply Cl/Cd loss only after epoch 20-30, when the model has established basic pressure patterns. Early Cl/Cd gradients may be noisy since the pressure field is still poorly predicted.
3. **Gradient clipping on Cl/Cd loss**: The integral produces diffuse gradients that could interact poorly with PCGrad. Clip the Cl/Cd gradient norm separately before adding to total loss.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2340 because ADVISOR: Closing after 2 iterations. The p_tan signal is real (-2.4% in v2, -2.9% in v1), but the fundamental issue is structural: the integral Cl/Cd gradient disrupts shared model weights for non-tandem predictions (p_in +2.9%, p_oodc +1.4%), and high seed variance suggests training instability. Neither tandem-only masking nor weight tuning will resolve this — the integral loss and per-node loss have competing gradient signals that PCGrad can't fully separate.

The good news: the insight that global aerodynamic integrals help tandem prediction is validated. But the right way to use it is as a *conditioning signal*, not a loss term. Fern: your next assignment will be a bold Round 40 experiment that builds on this insight architecturally..
